group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/EventDispatcher.java

```diff
@@ -0,0 +1,258 @@
+        return poll(Long.MAX_VALUE, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Returns the next {{@link Event}} available. This method blocks for the provided
+     * time and returns null of not event is available.
+     *
+     * @param timeout   The timeout.
+     * @param unit      The timeout unit.
+     * @return The next event available or null.
+     */
+    public T poll(long timeout, TimeUnit unit) {
+        lock.lock();
+        try {
+            K key = randomKey();
+            long nanos = unit.toNanos(timeout);
+            while (key == null && !closed && nanos > 0) {
+                try {
+                    nanos = condition.awaitNanos(nanos);
+                } catch (InterruptedException e) {
+                    // Ignore.
+                }
+                key = randomKey();
+            }
+
+            if (key == null) return null;
+
+            Queue<T> queue = queues.get(key);
+            T event = queue.poll();
+
+            if (queue.isEmpty()) queues.remove(key);
+            inflightKeys.add(key);
+            size--;
+
+            return event;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Marks the event as processed and releases the next event
+     * with the same key. This unblocks waiting threads.
+     *
+     * @param event The event that was processed.
+     */
+    public void done(T event) {
+        lock.lock();
+        try {
+            K key = event.key();
+            inflightKeys.remove(key);
+            if (queues.containsKey(key)) {
+                addAvailableKey(key);
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Returns the size of the queue.
+     *
+     * @return The size.
+     */
+    public int size() {
+        lock.lock();
+        try {
+            return size;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Closes the queue. This unblocks all the waiting threads.
+     */
+    @Override
+    public void close() {
+        lock.lock();
+        try {
+            closed = true;
+            condition.signalAll();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Adds the key to the available keys set.
+     *
+     * This method must be called while holding the lock.
+     */
+    private void addAvailableKey(K key) {
+        availableKeys.add(key);
+        condition.signalAll();
+    }
+
+    /**
+     * Returns the next available key. The key is selected randomly
+     * from the available keys set.
+     *
+     * This method must be called while holding the lock.
+     */
+    private K randomKey() {
+        if (availableKeys.isEmpty()) return null;
+
+        int lastIndex = availableKeys.size() - 1;
+        int randomIndex = random.nextInt(availableKeys.size());
+        K randomKey = availableKeys.get(randomIndex);
+        Collections.swap(availableKeys, randomIndex, lastIndex);
```

Comment:
Seems like the swapping here was done because this is an array list. What was the rationale for choosing this type?

Reply:
Yeah, I swap with the last in order to moving all elements in remove. I use an ArrayList because I need random access.